### DISCUSSION TITLE
Possible Bug: How do I configure the permissions to use relational data on field validations

### DISCUSSION THREAD

**Discussion by martinemmert** [directus/directus#6616] [Archive]
This might be something that is not possible or also might be an error.

# Background

## Idea

I want to create an upload hook, which moves the uploaded files into the user's subdirectory when a specific property is set on the uploaded files payload.

## Database alterations

I added a field to the `directus_folders` collection, named `owner`, a many-to-one relation between `directus_folders` and `directus_users.`

## Permission Configuration

The CURRENT_USER will belong to a role that has the following permission configuration:

### directus_folder

I configured the _read_ action as follows:

#### Item Permissions

```JSON
{
    "owner": "$CURRENT_USER"
}
```

#### Field Permissions

I granted access to the `name` and `owner` fields.

### directus_files

I configured the _update_ action as follows:

#### Item Permissions
```JSON
{
    "uploaded_by": "$CURRENT_USER"
}
```

#### Field Permissions:

granted access to the `folder` field

#### Field Validation:
```json
{
    "folder": {
        "owner": {
            "_eq": "$CURRENT_USER"
        }
    }
}
```

# Problem

The `files.upload` hook executes the following steps:

-> I try to get the folder of the current user
✅ this works

-> I create the folder of the current user if it does not exist
✅ this works

-> I try to update the file's folder field with the id of the current user's folder.
💥 This fails with the error `Invalid undefined schema.`

## Possible Error:

_I tracked the error message down to the following:_

Permission Object to check the access:

```JSON
{
  "permission": {
    "role": "5c3407a4-38f2-4f75-a872-1cc7d3417840",
    "collection": "directus_files",
    "action": "update",
    "permissions": {
      "uploaded_by": "$CURRENT_USER"
    },
    "validation": {
      "folder": {
        "owner": {
          "_eq": "$CURRENT_USER"
        }
      }
    },
    "presets": null,
    "fields": [
      "folder"
    ],
    "limit": null
  }
}
```

The [generateJoi](https://github.com/directus/directus/blob/18d6c0edbe4ccb35d4bd45fc6f5cde988e8e07c3/api/src/utils/generate-joi.ts#L56) method tries to use the `owner` as an operator, which will fail since there is no operator called `owner`

```js
// api/src/utils/generate-joi.ts:56
// I stripped away the typings to use it in a js file for debugging
function generateJoi(filter) {
	filter = filter || {};

	if (Object.keys(filter).length === 0) return Joi.any();

	let schema;

	for (const [key, value] of Object.entries(filter)) {
		if (key.startsWith('_') === false) {
			if (!schema) schema = {};

			const operator = Object.keys(value)[0];
			const val = Object.values(value)[0];

                         console.log({operator, val});
                         // { operator: 'owner', val: { _eq: 'test-owner' } }

			schema[key] = getJoi(operator, val);
		}
  }
  
        console.log("schema", schema);
        // schema { folder: undefined }

	return Joi.object(schema).unknown();
}
```

# Questions

A: Is it possible to validate with values of relations like this? 

or

B: is this an actual bug?

# Hook Source

```js
module.exports = ({ services }) => ({
  "files.upload": async ({ item, payload, schema, database, accountability }) => {

    if (!payload.contentUploadModule || !payload.contentUploadModule === "true") return;

    const { ItemsService, FoldersService, FilesService } = services;
    const serviceOptions = { schema, knex: database, accountability };
    const roomContentsService = new ItemsService("room_content", serviceOptions);

    // put file into directory of the user

    const foldersService = new FoldersService(serviceOptions);
    const filesService = new FilesService(serviceOptions);

    let folder;

    try {
      [folder] = await foldersService.readByQuery({
        filter: {
          owner: accountability.user,
          name: accountability.user,
        }
      });
      if (folder) folder = folder.id;
    } catch (error) {
      console.error(error);
    }

    if (!folder) {
      try {
        folder = await foldersService.createOne({
          owner: accountability.user,
          name: accountability.user,
        });
      } catch (error) {
        console.error(error);
      }
    }

    // TODO: add error number and put error into error list and what should happen if this error gets thrown?
    if (!folder) throw new Error("Could not create folder for user!");

    // move file into users folder
    try {
      await filesService.updateOne(item, { folder });
    } catch (error) {
      console.error(error);
    }
  },
});
```

**Comment by rijkvanzanten** [comment]
Heya!

It took me a couple reads to fully understand what was going on, but the problem here is the nested field under validation. The validation step currently doesn't handle nested fields yet: https://github.com/directus/directus/discussions/5086

**Reply by martinemmert** [reply]
Hey, thanks for replying.
Wow, I am bewildered that "nested object validation" did not come into my mind to describe this. Sorry for posting such a hard-to-read issue 😅.

**Reply by rijkvanzanten** [reply]
No worries!
